crates/solvers/src/domain/solver/dex/partial_fills.rs

```diff
@@ -0,0 +1,93 @@
+use {
+    crate::domain::{dex, eth, order},
+    std::{
+        collections::HashMap,
+        sync::Mutex,
+        time::{Duration, Instant},
+    },
+};
+
+/// Manages the search for a fillable amount for partially fillable orders.
+#[derive(Debug, Default)]
+pub struct PartialFiller {
+    /// Maps which fill amount should be tried next for a given order. For sell
+    /// orders the amount refers to the `sell` asset and for buy orders it
+    /// refers to the `buy` asset.
+    amounts: Mutex<HashMap<order::Uid, CacheEntry>>,
+}
+
+impl PartialFiller {
+    /// Returns which dex query should be tried for the given order. Takes
+    /// information of previous partial fill attempts into account.
+    pub fn dex_order(&self, order: &order::Order) -> dex::Order {
+        if !order.partially_fillable {
+            return dex::Order::new(order);
+        }
+
+        let (token, total_amount) = match order.side {
+            order::Side::Buy => (order.buy.token, order.buy.amount),
+            order::Side::Sell => (order.sell.token, order.sell.amount),
+        };
+        let now = Instant::now();
+
+        let amount = match self.amounts.lock().unwrap().entry(order.uid) {
```

Comment:
I can't remember now, but for some reason I thought locking for a `match` statement had some sharp edges.

Reply:
There are some sharp edges with this pattern:
```rust
match *mutex.lock().unwrap() {
    _ => {
        // long running code which does not need to have the mutex locked
    }
}
```

The lock is held for as long as the match statement is running which is a problem if we for example only want to clone the locked value and do some long running operation with it. In this case we need to keep the mutex locked for the entire match statement because we modify the value inside the mutex.